include/swift/SIL/PrunedLiveness.h

```diff
@@ -353,6 +353,33 @@ struct LiveRangeSummary {
 /// necessarily include liveness up to destroy_value or end_borrow
 /// instructions.
 class PrunedLiveness {
+public:
+  struct LifetimeEnding {
```

Comment:
I think the explanation for why certain kinds of uses override other uses belongs here now.

Reply:
Added this:
```
  /// A tristate describing how an instruction uses the value:
  /// - non-ending: the value's lifetime does not end at the instruction;
  ///               results from updateForUse(, lifetimeEnding: false)
  /// - ending: the value's lifetime ends at the instruction; results from
  ///           updateForUse(, lifetimeEnding: true)
  /// - non-use: the instruction doesn't use the value but liveness was extended
  ///            to it; results from extendToNonUse
  ///
  /// If an instruction is added to liveness with multiple LifetimeEnding
  /// instances, the stored instance needs to be updated appropriately, taking
  /// into account the instances that were already seen.
  ///
  /// For example, if Nonuse is seen after Ending, it is ignored: that the
  /// instruction ends the lifetime takes priority: the instruction doesn't end
  /// the lifetime of the value any less because liveness was extended to it.
  ///
  /// Similarly, if Ending is seen after NonEnding, it is ignored: that the
  /// instruction ends the lifetime of a copy of the value can't change the fact
  /// that liveness already extends _beyond_ the instruction because it was
  /// already recognized as a non-ending use.
  ///
  /// This relationship of "overriding" is captured by the order of the cases in
  /// LifetimeEnding::Value and which case overrides another is computed by
  /// taking the meet--i.e. the lesser of the two cases overrides.
  ///
  /// Note: Taking the meet may not be appropriate for branch instructions
  ///       which may need to be recognized as lifetime-ending when they have
  ///       as uses both a reborrow and a guaranteed phi.
```